### DISCUSSION TITLE
Importing the fast internals of boost-histogram

### DISCUSSION THREAD

**Discussion by TomNicholas** [scikit-hep/boost-histogram#586] [General]
I'm working on xhistogram, whose functionality we are planning to move upstream into xarray (https://github.com/pydata/xarray/pull/5400). 

Doing this meets a few requirements for us:
- xhistogram works with only numpy installed, optionally accelerated with dask, which is the same dependency structure of the rest of xarray's computation functions.
- xhistogram now scales very nicely to large datasets using dask (with any chunking pattern), and scaling is the main type of performance we wanted for our particular problems.
- xhistogram is "axis-aware" - the API allows the user to choose which dimension(s) they want to count over to create the histogram, broadcasting over the other axes automatically.

However, it would also be nice to be able to benefit from the seeming [faster performance](https://boost-histogram.readthedocs.io/en/latest/notebooks/xarray.html#Speed-comparson) of boost-histogram. Is it all possible for me to optionally import some part of boost-histogram and call it as a "fast path", if the user has the library installed? I don't think I can just use the high-level `Histogram` class for the various reasons above, but is there some lower-level part of the wrapped C++ API that would be useful to import (for bincounting say?).

**Comment by henryiii** [comment]
I would recommend (after reading https://boost-histogram.readthedocs.io/en/latest/notebooks/xarray.html  :) ) that you look at using [`boost_histogram.numpy`](https://boost-histogram.readthedocs.io/en/latest/api/boost_histogram.html#module-boost_histogram.numpy) - it provides the same API, but with speed up internals and optional threading; that should work well for making a "fast path" with minimal changes.

**Reply by henryiii** [reply]
Ah, though if you aren't using the NumPy histogram methods, but manually bin counting, etc, then the Histogram class is internally how this is implemented - the C++ Boost.Histogram's histogram class is the base for the computations.

**Reply by TomNicholas** [reply]
Thanks. Yes we are manually bincounting (partly to get the "axis-aware" behaviour I mentioned) but that's helpful.
